### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by Lancetnik** [airtai/faststream#1510]
# Roadmap

We know that some users are waiting for new broker support (like **SQS**/**MQTT**/etc.), but please be patient a bit longer.

Due to the inner project structure becoming much harder to change with each new broker, we are now focusing on stabilizing the inner and outer API. Then we will be able to promote new broker support in an easier way (like we used to do before).

### Polish Current Features

First of all, we should polish already existing brokers and core features:

* [ ] Complete Observability features
  * [ ] Add baggage feature support to tracing  (assigned to @draincoder)
  * [ ] Correct spans linking in batches case (assigned to @draincoder)
  * [ ] **Prometheus** Middleware to collect metrics
  * [ ] Healthcheck mechanism for k8s-probes (assigned to @Sehat1137)
* [ ] Strong Pseudo-random values generation (like message IDs)
  * [ ] Add Nuid or another pseudo-random algorithm implementation
  * [ ] Pass this object throughout the whole application to all places generating random values
* [ ] Extended **RMQ** features
  * [ ] Connections and channels pool support
  * [ ] Allow using a new channel for specific requests
  * [ ] Concurent RPC requests (using channel pool)
  * [ ] Publish Transactions support
* [ ] Extended **NATS** features
  * [ ] Create KeyValue adapter to provide custom methods
  * [ ] Create ObjectStore adapter to provide custom methods
  * [ ] In-memory KeyValue and ObjectStore tests
* [ ] Extended **Kafka** features
  * [ ] Transactions support
  * [ ] Pattern subscription
* [ ] Extended **Redis** support
  * [ ] Add Redis Pool support
  * [ ] Concurrent messages processing

### Inner Structure Refactoring

The next step is making `Context` non-global: we should pass it throughout the whole application from **FastStream** object -> Broker -> Subscriber -> Call. This step is also related to **FastDepends 3.0** migration, and we can refactor dependency_overrides as well, and finally make **Pydantic** replaceable in serialization cases.

* [ ] Optional Pydantic
  * [ ] Migrate to **FastDepends** 3.0
  * [ ] Provide an option to use custom Serializer Protocol implementation
* [ ] Local dependency-overrides
  * [ ] Migrate to **FastDepends** 3.0
  * [ ] Pass dependecy-overrides container from FastStream to the final consumers
  * [ ] Complete Depends documentation by overrides section
* [ ] Local context
  * [ ] Refactor all parsers to get regex in init instead of using context
  * [ ] Refactor `ContextType` to use local context
  * [ ] Pass context to **FastDepends** as kwargs
  * [ ] Pass context from `FastStream` to the final consumers
  * [ ] Refactor logging to not use raw context
* [ ] Finally, after all these steps, we can finalize the Middleware Public API
  * [ ] Use Response DTO instead of **options**
  * [ ] Add local context as an init option

### New in Testing

We plan to refactor the In-Memory `TestClient` to make it side-effect-less (currently, it has some problems with reusing).

* [ ] New TestClient logic
* [ ] Add tests for context

### New Brokers Support

Finally, we can take a break from refactoring and add new broker support:

* [ ] **SQS** support
* [ ] **MQTT** support

### 1.0.0 Finalization

At the end of the year (hopefully), we plan to release the first major version - 1.0.0

* [ ] Public Multi-broker support
  * [ ] **AsyncAPI** schema merging
  * [ ] Cross-brokers In-Memory Tests
  * [ ] Correct logging resolution
* [ ] Extended subscriber API
  * [ ] `broker.subsciber("test").get_one()` - get concrete messages support
  * [ ] `async for msg in broker.subscriber("test")` - messages iterator support

### Community

Additionally, we are planning to start our *Experts* activity and provide users with a roadmap to study **FastStream**, a form to approve your expertise, and a public page with all approved experts to encourage people to help each other (and boost their resumes with the **FastStream** expert mark).

* [ ] Launch **Experts** initiative

### AsyncAPI

Last but not least, we should migrate to the new major **AsyncAPI** version

* [ ] **AsyncAPI 3.0** support
* [ ] **AsyncAPI 2.6** compatibility

## 1.0.0 +

We are not looking too far ahead, but we have few thoughts:

* [ ] Enable users to publish a test message from **AsyncAPI** Web right to the app
  * [ ] With In-memory execution mode
  * [ ] With real broker publishing

Eventhough brokers are good, brokerless is much better, so **ZMQ** support looks promising to us

* [ ] **ZMQ** support

And the last thing - we want to provide you with some persistence layer (like **Faust** table or smth) to allow windowing incoming messages stream.

* [ ] Incoming stream windowing

P.S. This plan is still pretty raw, and we can swap any of these points, but this is the general direction.

**Comment by r0kk** [comment]
Hi thank you for sharing the roadmap, I would like to point out that community can already use MQTT protocol since RabbitMQ broker supports MQTT protocol via [plugin](https://www.rabbitmq.com/docs/mqtt). This means that FastStream can be used for with MQTT protocol. Maybe tutorial or announcement could inform community about this option.

Thank you for all contributions.

**Comment by Lancetnik** [comment]
@r0kk thank you for your kind words! But I am not sure, that plugin is full-MQTT compatible. Is it suitable for all MQTT cases? Did you test **FastStream** with it?
Anyway, we plan to implement MQTT support right after 0.6.0 release (it should be in a few months), it's not too long to wait

**Comment by matheusvnm** [comment]
@Lancetnik I just came across this project and found out that I'm currently doing something similiar for my company. We currently only work with PubSub via Google Platform with New Relic. @Lancetnik, do your team has the intention to provide support to Google PubSub? 

If so, I have a similar implementation for this "broker". In case you see some benefit in adding support for Google PubSub (and maybe New Relic optional instrumentation) I would like to contribute to the project.

**Comment by Lancetnik** [comment]
@matheusvnm, thank you for your interest and offer of help!

We are planning to introduce new broker support, and GCP is also on the plan - #1229.
However, my current priorities are MQTT and SQS.

Fortunately, we have an active contributor working on GCP support - #2444 . It seems like he has already done a lot of work, but I literally know nothing about GCP and am unable to deeply validate his work.

It would be great if you could coordinate and complete the task together - two heads (and pairs of hands) are much better than one. What do you think of this opportunity?

**Comment by rsJames-ttrpg** [comment]
@matheusvnm Hi, I am working on the gcp broker atm slowly progressing it. I could definitely do with a second pair of eyes if you want contribute. I don't have a huge amount of time at the moment but in a few weeks I should be able to dedicate actual time to it. In the mean time if you want to take a look at the pr and make sure it covers your use case / any missing gcp specific features, I have worked pretty extensively with pubsub but there are some features that I haven't touched and there are some of the gcp client features I have deliberately not implemented. If you want to chat I am on [discord](https://discord.com/users/201349522352373760)

**Comment by matheusvnm** [comment]
Hey, thanks for reaching out @rsJames-ttrpg! I'm definitely happy to contribute and collaborate on this.

I'm currently looking at the overall project to understand the main concepts and how it is structured. Once I do that I will
take a look at the PR to get familiar with your current progress. I can certainly check it against my use cases and provide feedback on any GCP-specific features that might be missing.

No worries about your time constraints; I too have a very time constrained schedule. I'll connect with you on Discord to chat more about the project on the weekend!

Thank you!

**Comment by devjerry0** [comment]
Hey Friends! checking in on the gcp progress here :)

**Comment by Lancetnik** [comment]
Hi @devjerry0! To be honest, GCP support isn't on our roadmap.

My current thinking for 1.0.0 is to keep a limited set of brokers supported in the core - just the most common ones: RMQ, Kafka, NATS, Redis, MQTT, SQS, and ZMQ. Support for other brokers (including GCP Pub/Sub) will be community-driven.
The project is maintained by a fairly small team, so we simply don't have the resources to support every broker ourselves. 

That's exactly why stable broker development APIs are one of the remaining items before 1.0 - once they land, building and maintaining a third-party broker integration should be much more approachable, and we'd be happy to point to community-maintained ones from the docs.

**Comment by Lancetnik** [comment]
We are planning to release **1.0** on the framework's birthday - **September 18th** 🎉

We have just a few things left to do:
* [x] Improve `TestClient` to be able to check all fields of incoming messages, not just the body - #557
* [ ] Improve the codec API - #2837
* [ ] Create a public API for third-party brokers - #2806
* [x] Migrate the FastAPI plugin to a separate repository - #2944

**Comment by hasansezertasan** [comment]
> * Migrate the FastAPI plugin to a separate repository

Working on Litestar Plugin for Faststream. I would recommend adding a page to the docs similar to https://starlette.dev/third-party-packages/

**Comment by Lancetnik** [comment]
A short status update, since this thread keeps getting the same question.

**Brokers in the core are a closed list**: **RabbitMQ**, **Kafka**, **NATS**, **Redis**, **MQTT**, **SQS** and **ZMQ**. Everything else — **GCP Pub/Sub**, **YDB**, **IBM MQ** and friends — is community-driven: a standalone package on top of **FastStream**'s public API. We are happy to help, review, and link it from the docs.

I've updated the description above to say this explicitly, so it no longer contradicts what I've been answering in the comments.

Community packages built this way already exist and are listed in the docs: **[Community Extensions](https://faststream.ag2.ai/latest/who-uses/#community-extensions)**. `stompman`, `faststream-mq`, `zMQTT`, `faststream-celery`, `litestar-faststream` and others are there. To add yours, open a PR to `docs/docs/en/who-uses.md`.

There is also a community-maintained list with a wider scope — libraries, tools, templates and resources: **[awesome-faststream](https://github.com/lesnik512/awesome-faststream)** by @lesnik512.

@hasansezertasan — that docs page is the one you suggested back in June, it's live now.

**Remaining before 1.0.0:**

* [ ] Improve the codec API — #2837
* [ ] Create a public API for third-party brokers — #2806

**On the date.** We originally aimed for September 18th, the framework's birthday. The scope of 1.0 has grown a bit along the way, so that date is moving — the release is in active development and should land in the near future. I'll update this thread as soon as we have a firm date.

The second one is what matters here: until it lands, a third-party broker has to build on internals that may still shift. That's why the answer to "can I add broker X?" right now is "wait for the release, then build it on the public API" — not "no".

Thanks to everyone who asked in this thread, it's what made the policy explicit.
